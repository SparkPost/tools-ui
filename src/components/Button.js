import React from 'react';
import { Link } from 'react-router';
import classNames from 'classnames';
const noop = () => {};

const mapPropsToClasses = ({ type, size, accent, fullWidth, icon, states, extraClasses }) => (
  classNames('button', {
    [`button--${type}`]: type,
    [`button--${size}`]: size,
    [`button--${accent}Accent`]: accent,
    'button--full': fullWidth,
    'button--icon': icon
  }, states, extraClasses)
);

const ActionButton = (props) => {
  const { action = noop, children } = props;
  const classes = mapPropsToClasses(props);

  return (
    <button onClick={action} className={classes}>
      {children}
    </button>
  );
};

const LinkButton = (props) => {
  const { to = null, children } = props;
  const classes = mapPropsToClasses(props);

  return (
    <Link to={to} className={classes}>
      {children}
    </Link>
  );
};

export { ActionButton, LinkButton };
