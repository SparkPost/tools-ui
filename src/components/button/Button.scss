@import "~@sparkpost/styles/src/assets/scss/shared/config";
@import "~@sparkpost/styles/src/assets/scss/shared/functions";

@import "~@sparkpost/styles/src/assets/scss/components/button";

.button {
  padding: 0 rem(12);
  white-space: nowrap;
}

.actionLink {
  display: inline-block;
  padding: 0 rem(6);

  color: color(orange);
  font-weight: 400;
  font-size: rem(14);

  &:hover, &:active, &:focus {
    cursor: pointer;
    text-decoration: none;
    color: color(orange--dark);
    background: grayscale(9);
  }

  @media screen and (min-width: break(md)) {
    padding: 0 rem(12);
    font-size: rem(15);
  }
}

.backLink {
  display: inline-block;
  margin-bottom: rem(9);
  font-size: rem(15);

  .fa {
    transition: 0.15s;
  }

  &:hover {
    text-decoration: none;

    .fa {
      transform: translate(-1px, 0);
    }
  }
}
