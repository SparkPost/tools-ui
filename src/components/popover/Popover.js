import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import Icon from '../../components/Icon';
import './Popover.scss';

// This is a popover that copies window.location.href
class CopyPopover extends Component {
  constructor(props) {
    super(props);
    this.state = {
      open: false
    };
  }

  componentDidMount() {
    document.addEventListener('click', (e) => this.handleClickOutside(e));
  }

  componentWillUnmount() {
    document.removeEventListener('click', (e) => this.handleClickOutside(e));
  }

  handleClickOutside(e) {
    const domNode = ReactDOM.findDOMNode(this);
    if ((!domNode || !domNode.contains(e.target))) {
      this.setState({ open: false });
    }
  }

  handleShareClick() {
    this.setState({ open: true });
    this.refs.urlToCopy.select();
    document.execCommand('copy');
    this.refs.urlToCopy.blur();
  }

  render() {
    const { children } = this.props;
    const { open } = this.state;

    return (
      <span className="popover__group" onClick={() => this.handleShareClick()}>
        {children}
        <div className={`popover copyPopover popover--l popover--bottom popover--rightAligned ${open && 'is-open'}`}>
          <h6 className='text--left copyPopover__label'><Icon name='check-circle' /> Copied to Clipboard!</h6>
          <input className='input__text input__text--s copyPopover__input'
            value={window.location.href}
            ref='urlToCopy' readOnly />
        </div>
      </span>
    );
  }
}

export { CopyPopover };
