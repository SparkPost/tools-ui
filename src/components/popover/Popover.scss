@import "~@sparkpost/styles/src/assets/scss/shared/config";
@import "~@sparkpost/styles/src/assets/scss/shared/functions";

.copyPopover {
  &__label {
    margin-bottom: 0;

    color: grayscale(1);
    font-family: $gotham-regular;
    font-size: rem(12);
    font-weight: 500;

    line-height: rem(18);

    .fa {
      display: inline-block;
      font-size: rem(15);
      color: color(green);
      width: rem(15);
    }
  }

  &__copiedString {
    padding: 0 0 0 rem(21);
    margin: 0;

    font-size: rem(12);
    color: grayscale(5);
    line-height: rem(15);
    text-align: left;
    word-break: break-all;

    &:hover, &:active, &:focus {
      border: none;
    }
  }

  &__hidden {
    position: absolute;
    opacity: 0;
    pointer-events: none;
  }

  &.popover.popover--l {
    width: 300px;
    @media screen and (min-width: break(md)) {
      width: 400px;
    }
  }
}
