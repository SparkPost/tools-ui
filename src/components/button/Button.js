import React from 'react';
import { Link } from 'react-router';
import Icon from 'components/Icon';
import { HoverPopover } from 'components/popover/Popover';
import config from 'config/index';
import getCurrentUrl from 'helpers/getCurrentUrl';
import classNames from 'classnames';
const noop = () => {};

import './Button.scss';

const mapPropsToClasses = ({ type, size, accent, fullWidth, icon, states, extraClasses }) => (
  classNames('button', {
    [`button--${type}`]: type,
    [`button--${size}`]: size,
    [`button--${accent}Accent`]: accent,
    'button--full': fullWidth,
    'button--icon': icon
  }, states, extraClasses)
);

const ActionButton = (props) => {
  const { action = noop, children } = props;
  const classes = mapPropsToClasses(props);

  return (
    <button onClick={action} className={classes}>
      {children}
    </button>
  );
};

const LinkButton = (props) => {
  const { to = null, children } = props;
  const classes = mapPropsToClasses(props);

  return (
    <Link to={to} className={classes}>
      {children}
    </Link>
  );
};

/**
 * Produces an orange link
 * Can be used to link or execute an action
 */
const ActionLink = (props) => {
  const { to = null, external = null, onClick = null, title = '', children } = props;

  if (external) {
    return <a href={external} className='actionLink' title={title}>{children}</a>;
  }

  return <Link to={to} onClick={onClick} className='actionLink' title={title}>{children}</Link>;
};

/**
 * Produces a blue link to indicate browser back
 */
const BackLink = ({ to = null, title = '' }) => <Link to={to} className='backLink' title={title}><Icon name='chevron-left' /> Back</Link>;

/**
 * Produces a Save Results action link
 */
const SaveResultsLink = () => {
  const currentUrl = getCurrentUrl(location);
  return (
    <HoverPopover placement='top' size='m' text='Create a free SparkPost account or login into your account to save results'>
      <ActionLink external={`${config.appUrl}/sign-up?return=${currentUrl}`} className='actionLink' title='Save Results'>Save Results</ActionLink>
    </HoverPopover>
  );
};

const SpLoginLink = ({ location = {}, classes, children }) => {
  const currentUrl = getCurrentUrl(location);
  const linkClasses = classNames('sp-sign-in', classes);
  return (
    <a href={`${config.appUrl}/auth?return=${currentUrl}`} title='Login' className={linkClasses}>{children}</a>
  );
};

const SpSignUpLink = ({ location = {}, classes, children }) => {
  const currentUrl = getCurrentUrl(location);
  const linkClasses = classNames('sp-sign-up', classes);
  return (
    <a href={`${config.appUrl}/sign-up?return=${currentUrl}&src=SP-Tools`} title='Sign Up' className={linkClasses}>{children}</a>
  );
};

export { ActionButton, LinkButton, ActionLink, BackLink, SpLoginLink, SpSignUpLink, SaveResultsLink };
