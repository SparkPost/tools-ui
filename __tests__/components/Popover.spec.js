import React from 'react';
import renderer from 'react-test-renderer';

import { CopyPopover } from 'components/popover/Popover';
import { ActionLink } from 'components/button/Button';

describe('CopyPopover component', () => {

  test('should render with only children prop', () => {
    expect(renderer.create(<CopyPopover><ActionLink>Click me!</ActionLink></CopyPopover>)).toMatchSnapshot();
  });

  test('should render when supplied a string to copy', () => {
    expect(renderer.create(<CopyPopover stringToCopy='Copy me!'><ActionLink>Click me!</ActionLink></CopyPopover>)).toMatchSnapshot();
  });

  test('should render when supplied a string to copy and styling props', () => {
    expect(renderer.create(<CopyPopover stringToCopy='Copy me!' block={true} placement='top'><ActionLink>Click me!</ActionLink></CopyPopover>)).toMatchSnapshot();
  });

});
