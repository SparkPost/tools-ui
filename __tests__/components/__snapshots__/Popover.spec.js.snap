exports[`CopyPopover component should render when supplied a string to copy 1`] = `
<span
  className="popover__group"
  onClick={[Function]}>
  <a
    className="actionLink"
    onClick={[Function]}
    style={Object {}}
    title="">
    Click me!
  </a>
  <div
    className="popover copyPopover popover--l popover--rightAligned popover--bottom">
    <h6
      className="text--left copyPopover__label">
      <i
        className="fa fa-check-circle " />
       Copied to Clipboard!
    </h6>
    <p
      className="copyPopover__copiedString">
      Copy me!
    </p>
    <input
      className="copyPopover__hidden"
      readOnly={true}
      value="Copy me!" />
  </div>
</span>
`;

exports[`CopyPopover component should render when supplied a string to copy and styling props 1`] = `
<span
  className="popover__group h-block"
  onClick={[Function]}>
  <a
    className="actionLink"
    onClick={[Function]}
    style={Object {}}
    title="">
    Click me!
  </a>
  <div
    className="popover copyPopover popover--l popover--rightAligned popover--top">
    <h6
      className="text--left copyPopover__label">
      <i
        className="fa fa-check-circle " />
       Copied to Clipboard!
    </h6>
    <p
      className="copyPopover__copiedString">
      Copy me!
    </p>
    <input
      className="copyPopover__hidden"
      readOnly={true}
      value="Copy me!" />
  </div>
</span>
`;

exports[`CopyPopover component should render with only children prop 1`] = `
<span
  className="popover__group"
  onClick={[Function]}>
  <a
    className="actionLink"
    onClick={[Function]}
    style={Object {}}
    title="">
    Click me!
  </a>
  <div
    className="popover copyPopover popover--l popover--rightAligned popover--bottom">
    <h6
      className="text--left copyPopover__label">
      <i
        className="fa fa-check-circle " />
       Copied to Clipboard!
    </h6>
    <p
      className="copyPopover__copiedString">
      http://localhost/
    </p>
    <input
      className="copyPopover__hidden"
      readOnly={true}
      value="http://localhost/" />
  </div>
</span>
`;
